src/librustc_infer/infer/combine.rs

```diff
@@ -164,7 +164,18 @@ impl<'infcx, 'tcx> InferCtxt<'infcx, 'tcx> {
             (_, ty::ConstKind::Infer(InferConst::Var(vid))) => {
                 return self.unify_const_variable(!a_is_expected, vid, a);
             }
-
+            (ty::ConstKind::Unevaluated(..), _)
+                if self.tcx.features().lazy_normalization_consts =>
+            {
+                relation.const_equate_obligation(a, b);
```

Comment:
So, I think that this logic should be

```rust
                // FIXME(#59490): Need to remove the leak check to accomodate
                // escaping bound variables here.
                if !a.has_escaping_bound_vars() && !b.has_escaping_bound_vars() {
                    relation.const_equate_obligation(a, b);
                }
                return Ok(b);
```

and the same for the line below. I created this test which ICEs with the branch as is:

```rust
// run-pass
#![feature(const_generics)]
#![feature(lazy_normalization_consts)]
#![allow(incomplete_features)]

// try to devise a test for higher-ranked case
//
// e.g., for<'a> EqZero<<() as Zero<'a>::C>>
//
//

trait EqZero<const C: usize> { }

impl EqZero<0> for () { }

trait Zero<'a> {
    const C: usize;
}

impl<'a> Zero<'a> for () {
    const C: usize = 0;
}

fn test_me<T>()
where for<'a> T: EqZero<{<() as Zero<'a>>::C}>
{
}

fn main() {
    test_me::<()>();
}
```

However, that test continues to ICE even with these changes, due to some assertion in type checking. So maybe it's fine, and we should just move that test to a follow-up issue. Anyway, this code as it is written here is the code we want *eventually* -- we just have to reconcile some details of the leak-check first.

Reply:
Added your suggestion. I don't know much about how HRTB are handled, so I think it might be better to deal with this afterwards.